File: packages/detached_membrane_sdk/detached_membrane_sdk/governance_e2e.py
"""Fixture-only governance e2e loop (no Ollama, no MacOS-CAS required)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .bhrt_projection import project_bhrt_packet
from .layered_verify import evaluate_layers
from .manifest import MANIFEST_REL_PATH, verify_manifest
from .pim0_emit import emit_pim0_from_proposal
from .proposal_emitter import emit_proposal_packet
from .verification_bridge import bridge_verification_result
from .ztna_local import issue_receipt, verify_receipt


GOVERNANCE_E2E_SCHEMA_VERSION = "detached-membrane-governance-e2e-0_1"
FIXTURE_REL = Path("fixtures/detached_membrane/governance_e2e_source_v0.json")
HOST_ACK_FIXTURE_REL = Path("fixtures/detached_membrane/macos_validator_ack_shape.json")


def _repo_root() -> Path:
    return Path(__file__).resolve().parents[3]


def _load_json(path: Path) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def run_governance_e2e(root: Path | None = None) -> dict[str, Any]:
    """Run fixture-only propose → ztna → layered verify → projection loop."""
    root = root or _repo_root()
    fixture = _load_json(root / FIXTURE_REL)
    host_ack = _load_json(root / HOST_ACK_FIXTURE_REL)

    manifest_ok, manifest_reasons = verify_manifest(root / MANIFEST_REL_PATH, root)
    if not manifest_ok:
        raise RuntimeError("manifest verify failed: " + "; ".join(manifest_reasons))

    source_packet_id = fixture["source_packet_id"]
    executor_profile_id = fixture.get("executor_profile_id", "python_agents_sdk")
    packet = emit_proposal_packet(
        source_packet_id=source_packet_id,
        executor_profile_id=executor_profile_id,
        summary=fixture.get("hint", "Governance e2e fixture proposal"),
    )

    policy_path = root / "packages/detached_membrane_sdk/policy/local_ztna_policy_v0.json"
    receipt_path = Path("/tmp/detached-membrane-governance-e2e-ztna.json")
    identity_ref = fixture.get("identity_ref", "detached_membrane_agent_local")
    ztna_receipt = issue_receipt(
        policy_path=policy_path,
        identity_ref=identity_ref,
        action="membrane_propose",
        resource="cas_return_packet",
        context_ref=source_packet_id,
        out_path=receipt_path,
    )
    ztna_verify_ok = verify_receipt(
        receipt=ztna_receipt,
        action="membrane_propose",
        resource="cas_return_packet",
        context_ref=source_packet_id,
    )

    packet["ztna"] = {
        "identity_ref": ztna_receipt.get("identity_ref"),
        "policy_decision_ref": ztna_receipt.get("decision_ref"),
        "decision_ttl_seconds": 900,
        "receipt_path": str(receipt_path),
        "authority_status": "advisory_only",
        "execution_permitted": False,
    }

    bridged = bridge_verification_result(packet=packet, host_ack=host_ack)
    layered = evaluate_layers(
        packet=packet,
        ztna=packet["ztna"],
        host_ack=host_ack,
        ztna_verify_ok=ztna_verify_ok,
        policy_assertions_ok=True,
        strict_legality=True,
    )
    pim0_envelope = emit_pim0_from_proposal(packet=packet)
    bhrt_projection = project_bhrt_packet(
        packet=packet,
        verification=bridged,
        ztna=packet["ztna"],
        pim0_envelope=pim0_envelope,
        parent_receipt_id=packet["ztna"].get("policy_decision_ref"),
    )

    return {
        "object": "DetachedMembraneGovernanceE2EReport",
        "schema_version": GOVERNANCE_E2E_SCHEMA_VERSION,
        "accepted": layered["accepted"] and bridged["accepted"] and manifest_ok,
        "manifest_reasons": manifest_reasons,
        "layered_verification": layered,
        "bridged_verification": bridged,
        "wyrm_trace_ref": bhrt_projection.get("wyrm_trace_ref"),
        "source_packet_id": source_packet_id,
        "fixture_mode": "no_ollama_no_macos_cas",
    }


def main() -> int:
    report = run_governance_e2e()
    print(json.dumps(report, indent=2))
    return 0 if report.get("accepted") else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/detached-membrane-governance/src/e2e/governance-e2e.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { projectBhrtPacket } from "../executor/bhrt-projection.ts";
import { emitPim0FromProposal } from "../executor/pim0-emit.ts";
import { emitProposalPacket } from "../executor/proposal-emit.ts";
import { bridgeVerificationResult } from "../executor/verification-bridge.ts";
import { evaluateLayers } from "../gates/layered-verify.ts";
import { verifyManifest } from "../manifest/index.ts";
import { MANIFEST_REL, REPO_ROOT, ZTNA_POLICY_REL } from "../paths.ts";
import { issueReceipt, verifyReceipt } from "../ztna/local.ts";

export const GOVERNANCE_E2E_SCHEMA_VERSION = "detached-membrane-governance-e2e-0_1";

export type GovernanceE2EReport = {
  object: "DetachedMembraneGovernanceE2EReport";
  schema_version: typeof GOVERNANCE_E2E_SCHEMA_VERSION;
  accepted: boolean;
  manifest_reasons: string[];
  layered_verification: ReturnType<typeof evaluateLayers>;
  bridged_verification: Record<string, unknown>;
  wyrm_trace_ref: string | null;
  source_packet_id: string;
  fixture_mode: "no_ollama_no_macos_cas";
};

export function runGovernanceE2e(root: string = REPO_ROOT): GovernanceE2EReport {
  const fixture = JSON.parse(
    readFileSync(join(root, "fixtures/detached_membrane/governance_e2e_source_v0.json"), "utf-8"),
  ) as Record<string, string>;
  const hostAck = JSON.parse(
    readFileSync(join(root, "fixtures/detached_membrane/macos_validator_ack_shape.json"), "utf-8"),
  ) as Record<string, unknown>;

  const { ok: manifestOk, reasons: manifestReasons } = verifyManifest(
    join(root, MANIFEST_REL),
    root,
  );
  if (!manifestOk) {
    throw new Error(`manifest verify failed: ${manifestReasons.join("; ")}`);
  }

  const sourcePacketId = fixture.source_packet_id;
  if (!sourcePacketId) {
    throw new Error("governance e2e fixture missing source_packet_id");
  }
  const executorProfileId = fixture.executor_profile_id ?? "python_agents_sdk";
  const packet = emitProposalPacket({
    sourcePacketId,
    executorProfileId,
    summary: fixture.hint ?? "Governance e2e fixture proposal",
    artifactId: "e2efixture0001",
  });

  const receiptPath = join("/tmp", "detached-membrane-governance-e2e-ztna.json");
  const identityRef = fixture.identity_ref ?? "detached_membrane_agent_local";
  const ztnaReceipt = issueReceipt({
    policyPath: join(root, ZTNA_POLICY_REL),
    identityRef,
    action: "membrane_propose",
    resource: "cas_return_packet",
    contextRef: sourcePacketId,
    outPath: receiptPath,
  });

  const ztnaVerifyOk = verifyReceipt({
    receipt: ztnaReceipt,
    action: "membrane_propose",
    resource: "cas_return_packet",
    contextRef: sourcePacketId,
  });

  const ztna = {
    identity_ref: ztnaReceipt.identity_ref,
    policy_decision_ref: ztnaReceipt.decision_ref,
    decision_ttl_seconds: 900,
    receipt_path: receiptPath,
    authority_status: "advisory_only",
    execution_permitted: false,
  };
  packet.ztna = ztna;

  const bridged = bridgeVerificationResult(packet, hostAck);
  const layered = evaluateLayers({
    packet,
    ztna,
    hostAck,
    ztnaVerifyOk,
    policyAssertionsOk: true,
    strictLegality: true,
  });

  const pim0Envelope = emitPim0FromProposal(packet);
  const bhrtProjection = projectBhrtPacket({
    packet,
    verification: bridged,
    ztna,
    pim0Envelope,
    parentReceiptId: ztna.policy_decision_ref,
  });

  return {
    object: "DetachedMembraneGovernanceE2EReport",
    schema_version: GOVERNANCE_E2E_SCHEMA_VERSION,
    accepted: layered.accepted && Boolean(bridged.accepted) && manifestOk,
    manifest_reasons: manifestReasons,
    layered_verification: layered,
    bridged_verification: bridged,
    wyrm_trace_ref: (bhrtProjection.wyrm_trace_ref as string | null) ?? null,
    source_packet_id: sourcePacketId,
    fixture_mode: "no_ollama_no_macos_cas",
  };
}
